File: remote/src/main/java/com/kelsos/mbrc/di/providers/ApiServiceProvider.java
package com.kelsos.mbrc.di.providers;

import com.google.inject.Inject;
import com.google.inject.Provider;

import retrofit2.Retrofit;

public class ApiServiceProvider<T> implements Provider<T> {
  @Inject private Retrofit retrofit;
  private Class<T> serviceClass;

  public ApiServiceProvider (Class<T> serviceClass) {
    this.serviceClass = serviceClass;
  }

  @Override
  public T get() {
    return retrofit.create(serviceClass);
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/di/providers/ApiServiceProvider.kt
package com.kelsos.mbrc.di.providers

import retrofit2.Retrofit
import javax.inject.Provider

abstract class ApiServiceProvider<T>
constructor(private val retrofit: Retrofit, private val classType: Class<T>) : Provider<T> {
  override fun get(): T {
    return retrofit.create(classType)
  }
}
